src/rpc/RPCClient.ts

```diff
@@ -133,45 +146,87 @@ class RPCClient<M extends ClientManifest> {
     return this.methodsProxy as MapCallers<M>;
   }
 
+  /**
+   * Generic caller for unary RPC calls.
+   * This returns the response in the provided type. No validation is done so
+   * make sure the types match the handler types.
+   * @param method - Method name of the RPC call
+   * @param parameters - Parameters to be provided with the RPC message. Matches
+   * the provided I type.
+   * @param ctx - ContextTimed used for timeouts and cancellation.
+   */
   @ready(new rpcErrors.ErrorRPCDestroyed())
   public async unaryCaller<I extends JSONValue, O extends JSONValue>(
     method: string,
     parameters: I,
+    ctx: Partial<ContextTimed> = {},
   ): Promise<O> {
-    const callerInterface = await this.duplexStreamCaller<I, O>(method);
+    const callerInterface = await this.duplexStreamCaller<I, O>(method, ctx);
     const reader = callerInterface.readable.getReader();
     const writer = callerInterface.writable.getWriter();
-    await writer.write(parameters);
-    const output = await reader.read();
-    if (output.done) {
-      throw new rpcErrors.ErrorRPCMissingResponse();
+    try {
+      await writer.write(parameters);
+      const output = await reader.read();
+      if (output.done) {
+        throw new rpcErrors.ErrorRPCMissingResponse();
+      }
+      await reader.cancel();
+      await writer.close();
+      return output.value;
+    } finally {
+      // Attempt clean up, ignore errors if already cleaned up
+      await reader.cancel().catch(() => {});
+      await writer.close().catch(() => {});
     }
-    await reader.cancel();
-    await writer.close();
-    return output.value;
   }
 
+  /**
+   * Generic caller for server streaming RPC calls.
+   * This returns a ReadableStream of the provided type. When finished, the
+   * readable needs to be cleaned up, otherwise cleanup happens mostly
+   * automatically.
+   * @param method - Method name of the RPC call
+   * @param parameters - Parameters to be provided with the RPC message. Matches
+   * the provided I type.
+   * @param ctx - ContextTimed used for timeouts and cancellation.
+   */
   @ready(new rpcErrors.ErrorRPCDestroyed())
   public async serverStreamCaller<I extends JSONValue, O extends JSONValue>(
     method: string,
     parameters: I,
+    ctx: Partial<ContextTimed> = {},
   ): Promise<ReadableStream<O>> {
-    const callerInterface = await this.duplexStreamCaller<I, O>(method);
+    const callerInterface = await this.duplexStreamCaller<I, O>(method, ctx);
     const writer = callerInterface.writable.getWriter();
-    await writer.write(parameters);
-    await writer.close();
-
-    return callerInterface.readable;
+    try {
+      await writer.write(parameters);
+      await writer.close();
+      return callerInterface.readable;
+    } finally {
+      // Clean up if any problems, ignore errors if already closed
+      await writer.close().catch(() => {});
```

Comment:
It's in a finally block so I wanted to avoid duplicating the error if there was one. In fact on reviewing it, i'm being dumb here, I should be cleaning up the readable if there is an error with the writable here. I'll fix it.

Reply:
I've fixed this up for the server streaming. Unary caller does the same thing however. 

Inside the try block we're closing the streams, these may throw errors but it's unlikely. In the finally we try the clean up but these will usually throw an error since the should already be closed in the try block. Any errors should come from inside the try block ,these are just here to double check that they closed in case of an error.

Maybe it could be done better?

Reply:
Every time there is an empty catch, we must have a comment explaining why. And it should point to later parts of the code which does either handle the exception for some reason.

Reply:
I've added comments for the `.catch()` usage.